src/zjs_callbacks.c

```diff
@@ -288,47 +271,42 @@ int32_t add_callback(jerry_value_t js_func,
     return new_cb->id;
 }
 
-int32_t zjs_add_callback(jerry_value_t js_func,
-                         jerry_value_t this,
-                         void* handle,
-                         zjs_post_callback_func post)
+zjs_callback_id zjs_add_callback(jerry_value_t js_func,
+                                 jerry_value_t this,
+                                 void* handle,
+                                 zjs_post_callback_func post)
 {
     return add_callback(js_func, this, handle, post, 0);
 }
 
-int32_t zjs_add_callback_once(jerry_value_t js_func,
-                              jerry_value_t this,
-                              void* handle,
-                              zjs_post_callback_func post)
+zjs_callback_id zjs_add_callback_once(jerry_value_t js_func,
+                                      jerry_value_t this,
+                                      void* handle,
+                                      zjs_post_callback_func post)
 {
     return add_callback(js_func, this, handle, post, 1);
 }
 
-void zjs_remove_callback(int32_t id)
+void zjs_remove_callback(zjs_callback_id id)
 {
     if (id != -1 && cb_map[id]) {
-        if (cb_map[id]->type == CALLBACK_TYPE_JS && cb_map[id]->js) {
-            if (cb_map[id]->js->func_list) {
+        if (GET_TYPE(cb_map[id]->flags) == CALLBACK_TYPE_JS) {
+            if (cb_map[id]->func_list) {
                 int i;
-                for (i = 0; i < cb_map[id]->js->num_funcs; ++i) {
-                    jerry_release_value(cb_map[id]->js->func_list[i]);
+                for (i = 0; i < cb_map[id]->num_funcs; ++i) {
+                    jerry_release_value(cb_map[id]->func_list[i]);
                 }
-                zjs_free(cb_map[id]->js->func_list);
-            } else {
-                jerry_release_value(cb_map[id]->js->js_func);
+                zjs_free(cb_map[id]->func_list);
```

Comment:
This looks wrong, you only get here if this is NULL. Instead you should be checking some kind of flag to see if you have saved a single function instead of a func_list and release that.